src/instance.rs

```diff
@@ -991,10 +991,52 @@ impl<T> std::fmt::Debug for Py<T> {
 pub type PyObject = Py<PyAny>;
 
 impl PyObject {
-    /// Casts the PyObject to a concrete Python object type.
+    /// Downcast this `PyObject` to a concrete Python type or pyclass.
+    ///
+    /// Note that you can often avoid downcasting yourself by just specifying
+    /// the desired type in function or method signatures.
+    /// However, manual downcasting is sometimes necessary.
+    ///
+    /// For extracting a Rust-only type, see [`Py::extract`](struct.Py.html#method.extract).
+    ///
+    /// # Example: Downcasting to a specific Python object
+    ///
+    /// ```rust
+    /// use pyo3::prelude::*;
+    /// use pyo3::types::{PyDict, PyList};
+    ///
+    /// Python::with_gil(|py| {
+    ///     let any: PyObject = PyDict::new(py).into();
+    ///
+    ///     assert!(any.downcast::<PyDict>(py).is_ok());
+    ///     assert!(any.downcast::<PyList>(py).is_err());
+    /// });
+    /// ```
+    ///
+    /// # Example: Getting a reference to a pyclass
+    ///
+    /// This is useful if you want to mutate a `PyObject` that
```

Comment:
> extracting PyRefMut<T>

This does a lot though; there is downcasting and then borrowchecking. I thought it better to show both steps explicitly.

Reply:
Indeed, which is why I suggesting showing both alternatives next to each other. The short-hand because once you understand what it is doing, it is more convenient. And the longer form which does explain it.

(If we have a general consensus against magic `FromPyObject` impls, we should probably remove that one. But if we have it, we should recommend it IMHO.)

Reply:
To be honest, I think examples are meant to contain code, not comments about code.

I would suggest just adding something like

```rust
// This combination is common, so that the following short-hand exists:
let class_ref: PyRefMut<Class> = class.extract()?;
assert_eq!(class_ref.i, 1);
```

Reply:
Done